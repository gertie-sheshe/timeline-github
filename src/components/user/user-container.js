import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import UserComponent from './user-component';
import Loader from '../loader/loader'
import HomePage from '../home-page/home-page';
import { fetchUser } from '../../redux/user/userActions';
import {selectSortedUser, selectError, selectHomePage, selectLoading, selectUser} from '../../redux/user/userSelectors';
import './user.scss'


class User extends Component {
  state = {
    user: ''
  }


  onChangeHandler = (e) => {
    e.preventDefault();

    this.setState({
      [e.target.name]: e.target.value
    })
  }

  onSubmit = (e) => {
    e.preventDefault();
    const { fetchUser } = this.props;
    const user = this.state.user;

    fetchUser(user);
  }

  render() {
    const { sortedUser, error, homePage, loading, user } = this.props;

    return (
      <div className="user-container">
        {!homePage && <h1>GitHub Timeline</h1>}
        {homePage && <HomePage/>}
        <form onSubmit={this.onSubmit}>
          <label />
          <input className="input" type="text" name="user" onChange={this.onChangeHandler} /><br/>
          <input className="button" type="submit" />
        </form>
        {loading && <Loader/>}
        {sortedUser && <UserComponent userRepos={sortedUser} />}
        {user && user.length === 0 && <p>{this.state.user} has no public repositories</p>}
        {error && <p>User not found</p>}
      </div>
    )
  }
}

const mapStateToProps = createStructuredSelector({
  sortedUser: selectSortedUser,
  error: selectError,
  homePage: selectHomePage,
  loading: selectLoading,
  user: selectUser
});

export default connect(mapStateToProps, {
  fetchUser
})(User);
