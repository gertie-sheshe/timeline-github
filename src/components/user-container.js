import React, { Component } from 'react';
import { connect } from 'react-redux';
import UserComponent from './user-component';
import './user.scss'

console.log('MAYAA', UserComponent)

class User extends Component {
  state = {}

  onChangeHandler = () => {
    console.log('on change handler')
  }

  onSubmit = (e) => {
    e.preventDefault();
    console.log('SUBMIT')
  }
  
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <label />
          <input type="text" onChange={this.onChangeHandler}/>
          <input type="submit"/>
        </form>
        <UserComponent />
      </div>
    )
  }
}

// const mapStateToProps = () => { }

// const mapDispatchToProps = () => { }

export default connect(null, null)(User);
