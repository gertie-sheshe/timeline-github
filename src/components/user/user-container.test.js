import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { User as UserContainer } from './user-container'

afterEach(cleanup);

describe('User Container', () => { 
    it('Should render User Container', () => { 
        const mockProps = {
            sortedUser: null,
            error: false,
            homePage: false,
            loading: false,
            user: null,
            fetchUser: jest.fn()
        };
        const { asFragment } = render(<UserContainer {...mockProps} />);
        expect(asFragment(<UserContainer/>)).toMatchSnapshot()
    });

    it('Should render error if present', () => {
        const mockProps = {
            sortedUser: null,
            error: true,
            homePage: false,
            loading: false,
            user: null,
            fetchUser: jest.fn()
        };

        const { getByTestId } = render(<UserContainer {...mockProps} />);
        expect(getByTestId('error')).toHaveTextContent('User not found');

    });

    it('Should render loader when set to true', () => {
        const mockProps = {
            sortedUser: null,
            error: false,
            homePage: false,
            loading: true,
            user: null,
            fetchUser: jest.fn()
        };

        const { getByTestId } = render(<UserContainer {...mockProps} />);
        expect(getByTestId('loader')).toBeInTheDocument()
    });

    it('Should fetch user', () => {
        const mockProps = {
            sortedUser: null,
            error: false,
            homePage: false,
            loading: false,
            user: null,
            fetchUser: jest.fn()
        };

        const { getByTestId } = render(<UserContainer {...mockProps} />);
        fireEvent.click(getByTestId('button'));
        expect(mockProps.fetchUser).toHaveBeenCalled()
    });

    it('Should give feedback if user has no repositories', () => {
        const mockProps = {
            sortedUser: null,
            error: false,
            homePage: false,
            loading: false,
            user: [],
            fetchUser: jest.fn()
        };

        const { getByTestId } = render(<UserContainer {...mockProps} />);
        expect(getByTestId('warning')).toHaveTextContent('Sorry, this user has no public repositories')
    })
});