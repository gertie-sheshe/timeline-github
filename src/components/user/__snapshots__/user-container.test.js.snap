// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Container Should render User Container 1`] = `
<DocumentFragment>
  <div
    class="user-container"
  >
    <h1>
      GitHub Timeline
    </h1>
    <form>
      <label
        for="username"
      >
        User:
      </label>
      <input
        class="input"
        id="username"
        name="user"
        type="text"
      />
      <br />
      <button
        class="button"
        data-testid="button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
